### ISSUE TITLE
Position sliders: When Distance not checked aria-valuetext not spoken consistantly

### ISSUE THREAD

**Issue by terracoda** [phetsims/gravity-force-lab-basics#113]
A new strategy for described position was implemented in #107.

It all works perfectly when Distance is checked. And the on-screen values in the A11y View for `aria-valuetext` are correct with or without Distance checked.

**In this issue we need to investigate:**
1. why the `aria-valuetext` is not being read out as expected when Distance is not checked,
2. determine if it is a VoiceOver-only issue like in #109 for the steppers, or if
3. this is an `aria-valuetext` issue caused by a non-change.

Because the `aria-valuetext` is working fine with Distance checked, my instinct is that this is NOT a VoiceOver-only issue.

**Possible reasons for the issue:**
- It's possible that old logic that was meant to silence region names except for on a region change is hanging around in the code somewhere. This code might be over in GFL Regular or in the shared code for these 2 sims.
- Or maybe because of item 3 above, the `aria-valuetext` (the actual text string) is not actually changing the browser and the AT do not speak it out. If this is the case, we will need to figure out some way around that. 

Ideally, what I would like to hear for position & distance is laid out in the grey rows of the 2 tables that represent the : [Cognitive Walkthrough for changing position](https://docs.google.com/document/d/1HdDG9ds2MdbCb21l9qk3cI8yBQxK6-wBWNXC4Tloji8/edit#heading=h.ty8avmawwxa).

**Specifically, with Distance not checked I'd like to hear:**
- On focus: described position & described distance (distance region name) **(already working)**
- First step & on directional changes: progress indicator & described distance
- Subsequent steps in the same direction: described distance (distance region name with or without a region name change)
- Upon hitting a max or min value for position: position landmark (last stop or track end) & described distance (distance region name) **(already working, but somewhat erratically)**
- Upon trying to go beyond a min or max value fro position: alert containing state information (describe distance & described force &/or value for force) **(already working)**

I'll try break up the tables so there is a set for when Distance is checked (currently white rows) and a set when Distance is not checked (currently grey rows). This should make the behaviour more clear and easier to verify.

Edited comment: added item 3.

**Comment by terracoda** [comment]
@zepumph, I'll unassign myself once I have the tables in the design doc updated.

**Comment by terracoda** [comment]
@zepumph, I have updated tables, but haven't split them up yet. Will do that before Monday.

**Comment by terracoda** [comment]
@zepumph, agreed that the this isssue is understandable without me splitting the table mentioned in https://github.com/phetsims/gravity-force-lab-basics/issues/113#issuecomment-479427805.

So I am going to not do that unless it is needed for more clarity later.

**Comment by zepumph** [comment]
When I had a listen with distance not checked, here is what I found based on the above bullets. Tested with NVDA on firefox.

>First step & on directional changes: progress indicator & described distance

To me this is working well, the "closer/farther" piece was always provided to me as expected.

>Subsequent steps in the same direction: described distance (distance region name with or without a region name change)

Here is I think the root of this github issue's problem. The value text for each subsequent movement is the same until the region change. So even though it is being reset on each slider movement, it is being set from "close to mass 1" to "close to mass 1" and as a result the screen reader won't read it out because it is thought to have not changed.

**Comment by zepumph** [comment]
I was able to work around this by adding code to the algorithm that detects this case, and adds an extra space to the end of the string, so the alerts oscillate between "close to mass 1" and "close to mass 1 "

This feels just fine, and I think I will keep it in here for now. This isn't a general piece for all aria value text setting, just in this particular position value text. The code looks basically like this:

```js

// get the aria value text we want

// if distance is not checked and the values are the same, add a space
        if ( !this.model.showDistanceProperty.get() && previousText === newAriaValueText ) {
          newAriaValueText = newAriaValueText + ' ';
        }

```

I'm actually happy with the simplicity of this edge case, and will make sure that we document this well.

**Comment by terracoda** [comment]
Thanks @zepumph for confirming that the `aria-valuetext` was not updating due to the lack of a change in the text. I was pretty sure that was the reason as the _quantitative distance_ `aria-valuetext` was working beautifully. 

And very cool! I was wondering if you would do something slick like add a space to trick the browser into to thinking like there was a change!

I'll do my verification in the morning!

**Comment by zepumph** [comment]
>Upon hitting a max or min value for position: position landmark (last stop or track end) & described distance (distance region name)

I also can't seem to find an issue with this bullet, perhaps the above issue was causing problems here too. @terracoda please confirm that this single fix has solved all the issues you were encountering with this bug.

**Comment by terracoda** [comment]
@zepumph, re https://github.com/phetsims/gravity-force-lab-basics/issues/113#issuecomment-481437969, it's probably fixed. I'll verify in morning :-)

**Comment by terracoda** [comment]
Described position and described distance are working beautifully in Master! 
In the sonification wrapper, it works well, too, but due to a KNOWN issue that might be a VoiceOver-only issue (no need to investigate here) I get duplicate force alerts (on everything), so sometimes the alerts drown out the `aria-valuetext`.

I'm closing this issue. Great work @zepumph!

**Comment by jessegreenberg** [comment]
I just found this issue during #156. If I understand correctly, alerts were missing because NVDA thought that alerts were identical and skipped the repeats. Previously, I had thought that we already had a workaround for this in ariaHerald 
```js
      // fully clear the old textContent so that sequential alerts with identical text will be announced, which
      // some screen readers might have prevented
      liveElement.textContent = '';
```
but perhaps this is not complete. @zepumph do you think this needs to be promoted to ariaHerald somehow?

**Comment by terracoda** [comment]
@jessegreenberg, good point. We will always need this kind of workaround for described values that span a range of actual slider values.

@zepumph could you comment the option in ariaHerald that already does this already?
@jessegreenberg, should this "clearing of the element" be a default option?

For current version of GFL Basics, everything is working as it should, so unassigning.

**Comment by terracoda** [comment]
Please re-assign me if you need me to test anything.

**Comment by jessegreenberg** [comment]
> should this "clearing of the element" be a default option?

Currently it is not optional, the content is always cleared. This was done to prevent this exact problem but this issue indicates that this workaround may not be sufficient anymore.

**Comment by zepumph** [comment]
Ugh yes @jessegreenberg, good catch. I can verify that the code in AriaHerald is not working for this case.

Here is the code that is the workaround in GFLBPositionDescriber:
```
        // // When distance isn't checked, the qualitative alerts are the same between region changes, so add a space such
        // // that the AT will still read the value text each time. See https://github.com/phetsims/gravity-force-lab-basics/issues/113#issuecomment-481413715
        if ( !this.showDistanceProperty.get() && previousText === newAriaValueText ) {
          newAriaValueText = newAriaValueText + ' ';
        }
```

To test, comment that if statement out, load the sim, uncheck the distance checkbox, and then move a mass position, notice that the valuetext is only spoken on NVDA on region change.

Note that I would LOVE to remove this workaround in GFLB. It is messy and some of the worst code I wrote while writing descriptions. 

I don't really know how to test araiHerald. Note that when I added two lines of code, nothing changed:

```js

      // fully clear the old textContent so that sequential alerts with identical text will be announced, which
      // some screen readers might have prevented
      liveElement.textContent = '';
      liveElement.textContent = Math.random();
      liveElement.textContent = '';

```

@jessegreenberg do you have any ideas?

**Comment by jessegreenberg** [comment]
Oh I am sorry, I misunderstood this issue completely, I thought this was related to aria-live alerts not aria-valuetext. The workaround in ariaHerald fixed this issue only for back-to-back duplicate aria-live alerts but wasn't related to aria-valuetext. In that case this is a new issue that we haven't seen yet. But it would still be good to move to common code somehow if possible.

**Comment by zepumph** [comment]
After discussing with @jessegreenberg, we decided that moving the "workaround," where we add a space to the end of the string, into AccessibleValueHandler was a good solution. In addition we did a bit of research and are using a hairspace instead of a normal space. We found that NVDA didn't say "space" when going over it char by char.

**Comment by zepumph** [comment]
From here I can go in and simplify the GFLBPositionDescriber code.

**Comment by zepumph** [comment]
Thanks @jessegreenberg for the help. I really like this change. Closing
